[NAS] 미션3: 체스판 초기화

Malloc72P opened the pull request:
안녕하세요?
백엔드 클래스의 NAS라고 해요! 😄 

미션3를 완료해서 PR을 남기게 되었어요! 🎺 

수정된 내용을 요약해보았습니다!

* 체스판을 구현하기 위해서, 체스판의 각각의 네모칸에 해당하는 ChessCell클래스를 생성했어요!
* 그리고 Board클래스에 ChessCell[][] 2차원 배열을 생성해서 체스판을 표현했어요.
* 셀 객체에 표현문자를 요청하면, 해당 셀에 체스말이 있으면 그 말의 표현문자를 리턴하고, 체스말이 없으면 .을 리턴하도록 구현했어요.
* 이번에는 최대한 접근제한자를 빼먹지 않도록 노력했어요! 잘 되었으면 좋겠네요. 😄 

이상으로 글을 마치겠습니다! 
요즘 날이 다시 추워졌는데 몸 조심하세요~! ❄️

Malloc72P committed:
BoardTest의 board변수에서 누락된 접근제한자를 추가하라

Malloc72P committed:
BoardTest의 람다표현식을 좀더 보기좋게 수정하라

- IntelliJ의 제안에 따라서 람다표현식을 수정함

Malloc72P committed:
폰을 표현하는 문자를 리턴하는 메서드를 테스트하는 코드를 작성하라

Malloc72P committed:
Pawn클래스에서 기물을 표현하는 문자를 리턴하는 메서드를 작성하라

- 색상에 따라 다른 문자를 리턴할 수 있도록, 멤버변수인 color를 참조해서 문자를 리턴하도록 구현함

Malloc72P committed:
체스판을 초기화하는 기능을 구현하라

- 체스판 구현을 위해서, 판에 있는 각각의 셀에 대응하는 클래스 ChessCell 클래스를 생성함
- 해당 셀에 어떤 체스말이 올라와 있는지는 occupiedBy멤버변수로 표현하도록 구현함
- 해당 셀의 상태(occupiedBy)를 이용해서 셀을 대표하는 문자(ex, p, P, `)를 리턴하는
  getRepresentation메서드를 정의함
- 체스판을 표현하기 위해서, Board 클래스에 ChessCell[8][8] 2차원 배열을 생성함
- 추가로 IntelliJ의 제안에 따라서, Board클래스의 생성자에서 ArrayList를 생성할 때,
  new ArrayList<>();로 생성하도록 수정함.

Malloc72P committed:
누락된 접근제한자를 추가하라

- Board클래스의 멤버변수 board에 접근제한자 private을 추가함

Malloc72P committed:
체스판 출력기능을 구현하라

- 출력기능 구현을 위해 StringBuilder를 사용해서 출력결과를 반환하는 메서드를 구현함

Malloc72P committed:
Main을 구현하라

- 게임의 시작과 종료기능 구현을 위해 Main클래스를 생성함.

Malloc72P committed:
ChessCell의 setOccupiedBy메서드에 대해 예외처리를 추가하라

- 이미 셀에 체스말이 들어가있는데 또 체스말을 넣으면 예외가 발생하도록 수정함
- 해당 예외상황을 테스트할 수 있도록 테스트클래스를 수정함
- 추가로 ChessCell에서 쓰지 않는 getter메서드를 제거함

LarryJung commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -1,15 +1,58 @@
+        }
+    }
+
+    public void print() {
+        System.out.println(getResultOfAllRows());
+    }
```
String을 리턴하기만하고 Main에서 리턴값을 출력하게 하는 것은 어떨까요?

Malloc72P replied:
사실 처음에는 print라는 메서드는 없고 getResultOfAllRows()라는 메서드만 있었어요. 
그래서 Main에서 이 메서드를 호출하고 String을 리턴받아서 출력하는 구조였어요.
그런데 요구사항을 보니까 print라는 메서드가 있더라구요. 

그래서 Board객체에 체스판을 콘솔창에 출력하는 기능(print메서드)까지 생기게 되었어요.

음.. Board객체에선 콘솔창에 출력하는 기능은 빼주는게 Board객체의 책임을 줄여주는데 도움이 될까요? 책임을 지게 하는것도 좋을 것 같고, 빼주는것도 좋을 것 같아서 고민되네요. 😄 

그런데 지금이라도 print메서드를 안쓰고 getResultOfAllRows()메서드를 Main에서 호출하면 출력결과에 해당하는 String을 리턴받을 수 있어요. 얘도 public으로 지정해놨거든요. 😄

LarryJung replied:
만약 이 체스게임이 진짜 ui가 들어가는 게임으로 만든다고 하면 Board는 출력을 할 수 있는 자료만 건네주고, 출력을 담당하는 쪽에서 print를 하든지.. 실제 ui로 그리던지 하겠죠..?

현재로는 이렇게 해도 저렇게 해도 복잡성이 늘어나지 않으니 괜찮기는 합니다~ ㅎㅎ 하지만 핵심 로직과, 입출력(IO) 기능은 보통 분리합니다.

LarryJung commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -1,15 +1,58 @@
 
 public class Board {
     private ArrayList<Pawn> pieceList;
+    private ChessCell[][] board;
```
Board 인스턴스도 board라는 이름을 사용할텐데, ChessCell[][] 는 chessCellMap, chessCells 처럼 이름짓는게 혼동이 덜 할 것 같네요~

Malloc72P replied:
앗 맞아요! 
저도 board라고 지어주면서도 좀 햇갈리는 이름같아서 거슬렸는데, 마땅한 이름이 생각나지를 않더라구요. 😞 
결국에는 그냥 board로 지었는데, 코멘트 달아주신걸 보니까 chessCells로 지어주면 혼동이 덜 할 것 같아요!
이 부분 수정해서 다시 커밋하겠습니다!

LarryJung commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -1,15 +1,58 @@
+import java.util.stream.Collectors;
 
 public class Board {
     private ArrayList<Pawn> pieceList;
```
이제 이 변수는 쓰지 않는 것 같으니 지워주세요~

Malloc72P replied:
오... 이 부분은 제가 너무 꼼꼼하지 못했네요...

처음엔 폰을 리스트로 관리하다가 체스판을 2차원 배열로 관리하는걸로 바꿨는데요, 
출력하는 기능만 구현하고 정작 이 부분이 바뀐건 생각하지 못했네요. 😞 

코멘트 달아주신 것처럼 pieceList변수를 지워야겠어요.
추가로 Board클래스의 findPawn()이랑 add()메서드, 그리고 size메서드까지 수정해줘야겠네요! 🧾
이 부분도 수정해서 다시 커밋하겠습니다!

LarryJung commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -1,15 +1,58 @@
+        return sb.toString();
+    }
+
+    private String getResultOfRow(int rowIdx) {
+        return Arrays.stream(board[rowIdx])
+                .map(cell -> String.valueOf(cell.getRepresentation()))
+                .collect(Collectors.joining());
+    }
```
오 스트림을 활용해서 잘 짜셨네요! 메서드를 활용하는 부분도 간결하게 눈에 잘 보여서 좋습니다. 👍

Malloc72P replied:
감사합니다! 다른 코드를 작성할때도 간결하게 짤 수 있도록 노력하겠습니다! 😄

LarryJung commented on src/main/java/net/nas/chess/ChessCell.java:
```diff
@@ -0,0 +1,20 @@
+
+public class ChessCell {
+    public static final char NONE_REPRESENTATION = '.';
+    private Pawn occupiedBy = null;
```
null 체크를 좀 더 신경쓰기 위해서? 일부러 null을 할당하신 느낌이네요. 
요럴때는 @Nullable 같은 어노테이션을 써주기도 합니다. 

사실 그렇지 않아도 null 체크는 늘 신경써야 할 부분입니다..

Malloc72P replied:
여기서 null을 사용한 이유는, 해당 셀에 할당된 체스말이 없다는걸 표현하기 위해서예요. 

대신 occupiedBy필드를 private으로 접근을 제한해서, 다른 클래스에서 해당 필드에 직접 접근하지 못하게 막았어요. 
괜히 접근해서 쓰다가 nullPointerException이 발생하면 안되니까요. 
대신 해당 셀에 있는 체스말의 정보를 획득하려고 한다면, ChessCell클래스의 public메서드를 사용하게 하려고 해요.
예를 들면, ChessCell클래스의 getRepresentation()메서드 처럼요.
```java
public char getRepresentation() {
    if (occupiedBy == null)
        return NONE_REPRESENTATION;
    else return occupiedBy.getRepresentation();
}
```
이렇게 해서, 해당 필드가 null이면, 비어있는 상태를 표현하는```.``` 문자를 리턴하구, 아니면 해당 필드에 있는 체스말을 리턴하도록 했어요. 
이렇게 하면 nullPointerException으로부터 안전하면서도 비어있는 셀을 표현할 수 있을 거라고 생각했어요! 😄 

그런데 한편으로는, 이렇게 구현하는게 과연 맞을까? 라는 생각이 들어요.
이렇게 해도, ChessCell클래스의 멤버메서드의 실수로 NullPointerException이 발생하면 큰일이니까요. 어떤 방법이 더 좋을까요? 이대로 가도 좋을까요? 아니면 다른 방법을 선택하는편이 더 좋을까요? 지금으로서는 제가 선택한 방법도 안전할 것 같다는 생각이 들어요! 😄

LarryJung replied:
지금은 멤버 변수와 사용하는 로직이 근처에 있어서`= null` 을 보고 어느정도 판단이 가능한데, 코드 덩치가 커지면 이 필드가 기본값이 null 이었을지 놓칠 수 있을거라 생각이 드네요~
그래서 원래는 필드를 사용할 때 꼼꼼하게 null 체크를 하는게 더 확실하고 좋습니다!
하지만 그렇다고 진짜 모든 필드를 전부 null체크를 하는게 수고스럽고, 하다 말다 하면 그게 더 버그를 유발할 수 있거든여

그래서.. 적용해볼 수 있는 것은
1. Optional 과
2. 디자인 패턴에서 Null Object Pattern 입니다.

`private Optional<Pawn> pawnOpt = Optional.empty();`
이렇게 선언하면 폰이 없을수도 있겠다라는 게 코드로써 파악이 가능하니 일종의 null체크를 강제로 하게 되는 장점이 있구요,

null object는 null 체크 로직을 위임할 객체를 대신 만드는 겁니다.
만약 enum에서 ColorOfChessPiece.BLANK 를 하나 더 만들고 체스셀을 null 대신에 blank pawn 으로 초기화 하면
체스 셀 안에 pawn은 모두 초기화가 되어 있고, null 이든 아니든 각 객체의 getRepresentation() 만 호출하면 되겠죠?
```java
public char getRepresentation() {
    return occupiedBy.getRepresentation();
}
```

LarryJung commented on src/test/java/net/nas/chess/BoardTest.java:
```diff
@@ -78,8 +78,23 @@ void testErrorAddition() {
     }
+
+    @Test
+    @DisplayName("2행은 흰색 폰, 7행은 검은색 폰만 있어야 합니다. 각각 pppppppp, PPPPPPPP로 출력되어야 합니다")
+    void testRepresentationsOfPawnLines() {
+        board.initialize();
+        assertThat(board.getWhitePawnsResult()).isEqualTo("pppppppp");
+        assertThat(board.getBlackPawnsResult()).isEqualTo("PPPPPPPP");
+    }
```
테스트의 내용은 DisplayName에 쓰여진 내용과 다르게 2, 7 행 테스트로 보이지는 않네요~
`@DisplayName("흰색 폰, 검은색 폰 행은 각각 pppppppp, PPPPPPPP로 출력되어야 합니다")`

정도가 테스트 내용과 조금 더 맞는 느낌이 드네요

Malloc72P replied:
아 이 부분은 제가 애매하게 DisplayName을 작성한 것 같아요. 😆 

여기에서 쓴 2행이랑 7행은 실제 체스판에서의 행을 표현하고자 하는거였어요.
근데 코드스쿼드 팀원들 덕분에, 체스판에서는 이런 행을 ```랭크```라고 표현한다는걸 알게 되었어요!

이 부분은 2행, 7행이라고 표현하기보다. 랭크2, 7 이렇게 표현하는게 더 좋을 것 같다는 생각이 들어요! 그러면 오해할 만한 부분이 줄어들 것 같아요! 😄 
이것도 수정해서 다시 커밋하겠습니다!

LarryJung commented on src/test/java/net/nas/chess/ChessCellTest.java:
```diff
@@ -0,0 +1,59 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class ChessCellTest {
```
꼼꼼하게 잘 작성해주셨네요.!

Malloc72P replied:
감사합니다! 다른 코드도 잘 작성하겠습니다! 😄

LarryJung commented on src/main/java/net/nas/chess/ChessCell.java:
```diff
@@ -0,0 +1,20 @@
+    public static final char NONE_REPRESENTATION = '.';
+    private Pawn occupiedBy = null;
+
+    public void setOccupiedBy(Pawn piece) {
+        if (occupiedBy != null)
+            throw new IllegalStateException("Chess Cell is already occupied by another piece");
+        occupiedBy = piece;
+    }
```
잘 하셨지만 좀 더 의견 드리면

pawn 멤버 변수는 pawn이나 piece로 네이밍하고
할당하는 메서드는 그냥 occupy 로 하는 것도 괜찮을 것 같은데 고려해보세요~

```java
public void occupy(Pawn pawn) { ... }
```

Malloc72P replied:
맞아요 occupiedBy랑 setOccupiedBy라고 하니까 뭔가 표현이 어색한 것 같아요.
저렇게 쓰면서도 어색하다는 생각이 들었는데, 딱히 대안이 떠오르지를 않더라구요 😞 

코멘트 달아주신것처럼 그냥 piece랑 occupy라고만 지어줘도 충분히 의미전달은 되면서, 덜 어색한 코드가 될 것 같아요. 그리고 주석도 달아주면 더 좋을 것 같네요!

이 부분도 수정해서 다시 커밋하겠습니다!

LarryJung commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -1,15 +1,58 @@
+
+    private String getResultOfRow(int rowIdx) {
+        return Arrays.stream(board[rowIdx])
+                .map(cell -> String.valueOf(cell.getRepresentation()))
```
Representation을 꺼내서 String으로 변환하는 경우가 잦아진다면 아예 getRepresentation의 리턴값을 String으로 변경하는 것도 고려해보세요~

Malloc72P replied:
이 부분이 좀 까다로운것 같네요. 😆 

lucas에 올라온걸 보니까 체스말을 문자로 표현할 때, char을 쓰더라구요.
그래서 아래와 같이 상수값을 선언해놨어요.
```java
public static final char WHITE_REPRESENTATION = 'p';
public static final char BLACK_REPRESENTATION = 'P';
```
그렇다보니, getRepresentation()메서드를 구현하면 저 상수값을 쓰게되니까 어쩔 수 없이 String에서 char로 변환하는 일이 발생하게 되었네요. 🚢 

음... 이 부분은 어떻게 처리하는게 좋을까요?
제 생각에는 그냥 상수값 자체의 타입을 char에서 String으로  바꾸는게 좋을 것 같다는 생각이 들어요. 굳이 이 상수값이 char일 필요는 없을 것 같아서요.

그런데 한편으로는 이렇게 바꾸면 요구사항을 어기는게 아닌가 하는 생각이 들어서 고민이 되네요. 아예 char타입의 상수값이랑 String타입의 상수값을 만들어두고 쓰는건 어떨까요? 타입만 다르고, 값은 다른거예요. 아래처럼요.
```java
public static final char CHAR_WHITE_REPRESENTATION = 'p';
public static final String STRING_WHITE_REPRESENTATION = "p";
```
이 방법도 좋을 것 같은데, 리뷰어님은 어떻게 생각하세요?

LarryJung replied:
네 저는 char -> String으로 바꾼다고 해서 요구사항을 크게 벗어난다고 생각하지 않습니다~ 
char로 해서 코드 중복이 발생하고 보기가 싫어진다..(?) 라고 생각하면 String으로 짜도 되고, 
String으로 했다가 어떤 이유에서 char가 더 좋을것 같아라고 생각이 들면 또 바꾸면 되죠 ㅎㅎ

LarryJung reviewed (commented):
수고하셨습니다~ 요구사항을 깔끔하게 잘 작성해 주셨네요. 조금 더 신경써볼 부분들을 리뷰로 남겨드렸으니 확인해주세요.^^

Malloc72P committed:
Board클래스의 멤버변수 board의 이름을 혼동이 덜한 이름으로 바꿔라

- board로 이름을 지어주면, Board의 인스턴스가 될 수 있는 board랑 혼동이 올 수 있음.
- 그래서 board 대신, chessCells로 바꿔주었음.

Malloc72P committed:
테스트메서드의 DisplayName을 오해의 소지가 없도록 수정하라

- 2행, 7행이라고 표현했는데, 실제 배열에서는 1행, 6행이라 오해의 소지가 있음.
- 그래서 체스용어인 랭크2, 랭크7로 수정하여, 체스판에서의 좌표로 읽을 수 있도록 수정함

Malloc72P committed:
ChessCell의 멤버변수와 메서드명을 덜 어색하게 수정해라

- occupiedBy랑 setOccupiedBy는 좀 어색했음
- 그래서 occupiedBy는 piece로, setOccupiedBy는 occupy로 수정하여 어색한 부분을 줄이려는 노력을 하였음.
- 추가로, 해당 메서드명과 관계있는 테스트메서드의 이름도 수정하였음.

Malloc72P committed:
매직넘버를 줄여라

- Board클래스에서 흰색, 검은색 폰의 랭크를 1, 6으로 쓰고 있었음.
- 코드만 봐서는 1, 6번 인덱스가 무엇을 의미하는지 알기 어려웠음.
- 그래서 상수값으로 만들어서 상수의 의미를 알기 쉽게해서, 코드를 이해하기 쉽게 만드려는 노력을 하였음.

Malloc72P committed:
안쓰는 변수 pieceList를 지우고, add, find메서드를 고쳐라

- 기존의 add, find메서드는 더이상 안쓰는 변수인 pieceList를 쓰는 상태였음
- pieceList를 더이상 안쓰는 관계로 지웠음.
- 그래서 add, find메서드가 ChessCell[][]변수인 chessCells를 쓰도록 수정함
- 해당 변경으로 인해, 테스트 메서드도 수정하고, 다시 테스트하였음.

Malloc72P committed:
Board클래스의 initialize메서드에서도 add()메서드를 쓰도록 수정하라

- add메서드로 보드에 체스말을 추가하면, numberOfPieces가 증가함
- 그런데 기존의 initialize에선 add메서드도 안쓰고, numberOfPieces변수도
  수정하지 않아서, 보드에 체스말은 배치되었으나, size는 여전히 0인 문제가 있었음.
- 그래서 initialize메서드에서도 add메서드를 쓰도록 수정함
- 추가로 테스트메서드에서도 초기화 후에 크기가 16인지 검사하도록 수정함.
  아직까지는 폰만 추가한거라 초기화 후엔 16개가 배치되어있기 때문에 16임.

Malloc72P commented:
안녕하세요? 
코멘트해주신 내용을 참고하여 코드를 추가로 수정했어요! 😄 

수정된 내용을 요약하면 아래와 같습니다!
#### 애매모호하거나 어색한 이름을 수정했습니다!
* Board클래스에서 board라는 멤버변수를 chessCells로 수정했어요!
* ChessCell클래스에서 occupiedBy라는 멤버변수와 setOccupiedBy메서드를 각각 piece, occupy로 수정했습니다!
* Board클래스의 테스트메서드에서 오해의 소지가 있던 메서드의 이름과 DisplayName을 수정했어요. testRepresentationsOfPawnLines에서 testInitialization()으로 수정했어요. 초기화 메서드가 잘 동작하는지 확인하는거니까, 새로운 이름이 더 맞는 것 같네요! 그리고 DisplayName도 수정했어요!
* 변경 전
```java
@DisplayName("2행은 흰색 폰, 7행은 검은색 폰만 있어야 합니다. 각각 pppppppp, PPPPPPPP로 출력되어야 합니다")
void testRepresentationsOfPawnLines()
```
* 변경 후
```java
@DisplayName("초기화 후에는 흰색폰과 검정색폰이 각각 랭크 2, 7에 한줄씩 배치되어야 합니다.")
void testInitialization()
```
#### 안쓰는 변수인 Board클래스의 pieceList를 삭제했어요!
* 해당 수정으로 인해, Board클래스의 add, findPawn, size메서드가 수정되었어요. 해당 수정에 영향을 받은 테스트메서드도 새로 수정하고 다시 테스트하였습니다!
#### 매직넘버대신 상수값을 쓰도록 수정했습니다
* 코드에서 2, 6, 8 이런 값을 쓰기 보다, 아래와 같이 상수값을 선언하고 쓰도록 해서 코드의 이해를 높이려고 했어요
```java
public static final int LENGTH_OF_BOARD = 8;
public static final int RANK_OF_WHITE_PAWNS = 2;
public static final int RANK_OF_BLACK_PAWNS = 7;
```

Malloc72P committed:
체스말을 표현하는 문자상수의 타입을 String으로 바꿔라

- 기존에는 char이었는데, 정작 쓸때는 String타입으로 바꿔서 사용했음.
- 그래서 상수의 타입을 String 변경하기로 결정함.
- 해당 변경에 영향이 가는 테스트메서드도 수정함.

ghojeong commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -1,30 +1,90 @@
+ */
 public class Board {
-    private ArrayList<Pawn> pieceList;
+    public static final int LENGTH_OF_BOARD = 8;
+    public static final int RANK_OF_WHITE_PAWNS = 2;
+    public static final int RANK_OF_BLACK_PAWNS = 7;
```
상수를 따로 정의하신것 인상 깊네요.
저도 Row 라는 표현을 쓸지 Rank 라는 표현을 쓸지 고민을 했었는데,
저는 체스를 잘 모르는 사람도 이해하기 쉽도록 Row 라는 용어를 선택했습니다.
하지만 체스 도메인 입장에서는 Rank 라는 표현이 더 적절할 것 같기도 하네요.

Malloc72P replied:
저도 체스는 잘 몰라서 rank라는 이름이 잘 와닿지는 않네요 ㅎㅎ
Row라는 이름은 누구나 이해하기 쉬운 이름이니까 이것도 좋을 것 같아요! 👍

ghojeong commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -1,30 +1,90 @@
+    private int numberOfPieces = 0;
 
     public Board() {
-        pieceList = new ArrayList<Pawn>();
+        chessCells = new ChessCell[8][8];
```
오! Cell 이라는 네이밍 매우 마음에 듭니다!
저는 저걸 뭐라고 표현해야하나 하다가, Square 라는 말을 써버렸네요

ghojeong commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -1,30 +1,90 @@
+    public void add(Pawn piece, int rankIdx, int fileIdx) {
         if (piece == null)
-            throw new InvalidParameterException("Null value cannot be added in board");
-        pieceList.add(piece);
+            throw new InvalidParameterException("Null value cannot be added in Board");
+        if (isInvalidIdx(rankIdx) || isInvalidIdx(fileIdx))
+            throw new InvalidParameterException("index exceeded the bounds of the Board");
```
와! 세심한 예외처리 인상깊습니다.
`java.security.InvalidParameterException;`
을 사용하셨는데, RuntimeException 을 상속받아서 NAS 만의 커스텀 Exception 을 만드시는건 어떠실지 제안해봅니다.
지금은 많지 않지만, 나중에 에러 메시지만으로 에러를 구분하려고 하면 헷갈릴수도 있을 것 같아서요.

Malloc72P replied:
오... 저만의 커스텀 익셉션이라니 멋있네요 ㅋㅋ 
이것도 적용해보면 좋을 것 같아요! 공부 겸 햇갈림 방지 겸 ㅋㅋ
이건 미션 4 진행하면서 추가해보면 좋을 것 같아요! 그때 한번 도전해보겠습니다!

ghojeong commented on src/main/java/net/nas/chess/ChessCell.java:
```diff
@@ -0,0 +1,24 @@
+
+public class ChessCell {
+    public static final char NONE_REPRESENTATION = '.';
+    private Pawn piece = null;
```
Piece 인터페이스를 새로 하나 만드시는 건 어떠실지 제안해봅니다.

ghojeong commented on src/main/java/net/nas/core/Main.java:
```diff
@@ -0,0 +1,49 @@
+        System.out.print("게임을 시작하려면 start를 입력해주세요\n> ");
+        while (!isTerminated) {
+            String decision = scan.nextLine();
+            switch (decision.toUpperCase()) {
```
대소문자 상관없이 입력값이 실행되게 하는 세심한 배려 좋습니다. 👍

ghojeong commented on src/test/java/net/nas/chess/ChessCellTest.java:
```diff
@@ -0,0 +1,59 @@
+        assertThatThrownBy(() -> cell.occupy((Pawn) testcase))
+                .isInstanceOf(exceptionClass);
+    }
+}
```
EOF

Malloc72P replied:
앗 이부분을 놓쳤네요 ㅎㅎ. 이 부분도 커밋해서 푸시하겠습니다! 😆 
고마워요 파이로!

Malloc72P committed:
더이상 안쓰는 ChessCell클래스를 제거하라

- ChessCell클래스는 체스판 위의 네모칸을 표현하는 클래스이다.
- 이 클래스의 역할은 비어있는 네모칸을 처리할 때, null처리를 안전하게 해주는것이다.
- 하지만, ChessCell클래스에 배치된 체스말을 획득하는 getter메서드를 쓰면, 결국 null값이 리턴되는 문제가 있었다.
- 이 문제를 해결하기 위해, 리뷰어님과 코쿼팀원분들의 조언을 구했다.
- Color에 BLANK라는 상수를 추가하고, 이걸 Pawn객체에 넣어서 빈칸을 구현하기로 했다.
- 덕분에 Board클래스에서 null을 안써도 빈칸을 구현할 수 있게 되었다.
- 그래서 ChessCell클래스는 필요가 없어졌다. null값이 없으니, ChessCell도 필요가 없어졌기 때문이다.
- 추가로 위의 변경에 영향받은 코드도 수정하였다.
- ChessCell클래스와 해당 클래스의 테스트 클래스를 제거하였다.

Malloc72P committed:
Print메서드를 Board에서 Main으로 옮겨라

- 기존에는 Board클래스에서 print까지 했는데, 앞으로를 위해서 해당 기능을 Main클래스로 옮김
- Board클래스에서는 Board의 상태를 문자열로 리턴하는것만 하도록 수정함

Malloc72P committed:
getResultOfRow메서드명을 이해하기 쉬우면서 일관되도록 수정하라

- 다른 코드에서는 rank라는 이름을 쓰고 있으므로, 일관성있게 rank를 쓰도록 수정함
- 추가로, rank를 표현하는 문자열을 리턴하는거니까, getResult대신, getRepresentation으로 수정함

Malloc72P commented:
안녕하세요!
코멘트 해주신 내용을 참고하여 코드를 추가로 수정했어요! 😎

수정된 내용은 아래와 같습니다!
### 1. [null 처리](https://github.com/codesquad-members-2021/java-chess/pull/87#discussion_r577546609)를 위해서 코드를 수정했어요!
코멘트해주신 내용 잘 읽었어요! 👍 추가로 팀원들이랑 어떻게 하면 좋을지 고민해보았습니다!
그래서 Color에 BLANK를 추가하고, 비어있는 체스말을 표현할 수 있도록 Pawn클래스를 수정했어요. 그리고 Board에서 초기화할때, 8 X 8 2차원 배열을 만들고, 각각의 요소마다 Pawn객체를 생성해서 넣었어요! 다만, 비어있는 칸에 넣는 Pawn객체는 color를 BLANK로 해서, 체스 보드에서 비어있는 칸을 표현했어요! 이렇게 하니까 null값이 생길일이 없어서 nullPointerException의 위험이 많이 줄어든 것 같아요!
이렇게 바꾸고 나니까 더 이상 ChessCell클래스는 필요없어서, 테스트클래스와 함께 제거했어요. 그리고 Board클래스에서 `ChessCell[][] chessCells`를 `Pawn[][] chessCells`로 수정했어요!

### 2. [Print의 책임](https://github.com/codesquad-members-2021/java-chess/pull/87#discussion_r577540558)을 Board에서 Main으로 옮겼어요!
나중을 생각해봐도 그렇고, 당장을 봐도 굳이 Board에서 출력까지 할 필요는 없을 것 같다는 생각이 들었어요. 그래서 Board클래스는 보드의 상태를 알려주는 책임만 갖도록 수정했어요. 대신 Main클래스에서는 게임을 진행하는 책임이 있으니까, Board의 상태를 읽어와서 화면에 출력하는 역할을 부여했어요. 그리고 메서드 이름을 보고 기능을 좀 더 쉽게 알 수 있도록, `getResultOfAllRows()`에서 `getRepresentationOfBoard()`로 수정했어요!

### 3. [char 또는 String](https://github.com/codesquad-members-2021/java-chess/pull/87#discussion_r577547980)으로 상수를 표현할 지 고민하던 문제
상수값의 타입을 String으로 바꾸기로 했어요!
현재, 상수타입이 char타입이어야 하는 코드는 없는데, String타입이어야 하는 부분은 있더라구요. 그래서 String타입으로 수정했어요! 만약 나중에 char타입이 아니라서 문제가 생긴다면, 그때가서 수정하면 될 것 같아요! 😊

이상으로 요약을 마칠께요! 
제 코드를 리뷰해주셔서 정말 감사합니다! 😊

honux77 commented:
👍 @LarryJung 요거는 제가 머지할게요! 수고하셨습니다.
